grub-core/loader/arm64/linux.c

```diff
@@ -389,7 +391,10 @@ grub_cmd_linux (grub_command_t cmd __attribute__ ((unused)),
   grub_loader_unset();
 
   kernel_alloc_pages = GRUB_EFI_BYTES_TO_PAGES (kernel_size + align - 1);
-  kernel_alloc_addr = grub_efi_allocate_any_pages (kernel_alloc_pages);
+  kernel_alloc_addr = grub_efi_allocate_pages_real (GRUB_EFI_MAX_USABLE_ADDRESS,
```

Comment:
I wonder if isn't better to instead make `grub_efi_allocate_any_pages()` to use `GRUB_EFI_LOADER_CODE` rather than open coding it the params that use it to call `grub_efi_allocate_pages_real()`.

Since AFAICT is only used to allocate memory for the kernel image anyways, the memory for the initrd (that's data and not code) is already allocated using a different helper `allocate_initrd_mem()` so that can continue to use `GRUB_EFI_LOADER_DATA`.

Reply:
Also, by changing `grub_efi_allocate_any_pages()` you will fix this not only for arm64 but also for x86.

Reply:
It seems that it's used in [`grub_efi_mm_init()`](https://github.com/rhboot/grub2/blob/b1ce9c4d9da35364d54adbceeee9474a57ed80ad/grub-core/kern/efi/mm.c#L704) to allocate memory maps. How about a helper (perhaps `grub_efi_allocate_executable_pages()` or `grub_efi_allocate_kernel_pages()`)?

Reply:
I see, probably `grub_efi_allocate_kernel_pages()` then to be consistent with how the memory for the initrd is allocated ?

Reply:
i386 already has https://github.com/rhboot/grub2/commit/8a2d168012a6b6a2a636a9d50cea1d0f4c999ba5 (although that uses `EFI_RUNTIME_SERVICES_CODE` instead). I assume that's also used for x86_64?

Reply:
oh, that's actually a good question. I don't know the difference between `GRUB_EFI_RUNTIME_SERVICES_CODE` and `GRUB_EFI_LOADER_CODE`. Maybe the former should be used in this case too? I'll let @vathpela to advise here.

Reply:
I honestly have no clue here. I'm just going off of table 29 / page 159 in https://uefi.org/sites/default/files/resources/UEFI_Spec_2_8_final.pdf:

> `EfiLoaderCode`: The code portions of a loaded UEFI application.

> `EfiRuntimeServicesCode`: The code portions of a loaded UEFI Runtime Driver.

To me, the EFI stub of the kernel seems more like a UEFI application than a runtime driver, but again, I don't know much about this stuff.

Reply:
> To me, the EFI stub of the kernel seems more like a UEFI application than a runtime driver, but again, I don't know much about this stuff.

Yes, I agree with you here.

Reply:
Fwiw: chainloader also seems to use `GRUB_EFI_LOADER_CORE`: https://github.com/rhboot/grub2/blob/b1ce9c4d9da35364d54adbceeee9474a57ed80ad/grub-core/loader/efi/chainloader.c#L1052-L1054

Reply:
> It seems that it's used in [`grub_efi_mm_init()`](https://github.com/rhboot/grub2/blob/b1ce9c4d9da35364d54adbceeee9474a57ed80ad/grub-core/kern/efi/mm.c#L704) to allocate memory maps. How about a helper (perhaps `grub_efi_allocate_executable_pages()` or `grub_efi_allocate_kernel_pages()`)?

Probably just move kernel_alloc() from the i386 loader to some common code and make the policy list be passed in?

Reply:
That seems to depend on some interaction with the `max_addresses` field, where parts of it are read in from the i386/x64 kernel header. I'm not sure how to adapt this for AArch64, but if you have some patches I'd be happy to test.

Reply:
> Probably just move kernel_alloc() from the i386 loader to some common code and make the policy list be passed in?

So after having another look at this, the `max_addresses`  thing shouldn't be an issue. As you've said, this could just be passed in. However, I think using `kernel_alloc()` here doesn't make that much sense to do this in this PR:

We could use it for the kernel image + command-line memory, passing in a policy array with a single entry. But given that it only uses a single entry and none of the re-try logic that's used on x86, this seems a bit too convoluted to me. Also, it's orthogonal to the change to `GRUB_EFI_LOADER_CODE`, because that is still passed in to `kernel_alloc()`. Further, adapting it for the initrd would need some changes (or workarounds) as the arm64 code relies on `grub_efi_get_ram_base()` whereas the x86 code doesn't.